Update CalculiX adapter documentation (#180)
precice/precice.github.io: pages/docs/adapters/calculix/adapter-calculix-configure.md

precice-config-file: ../precice-config.xml
```

> The adapter allows to use several participants in one simulation (e.g. several instances of Calculix if several solid objects are taken into account). The name of the participant "Calculix" must match the specification of the participant on the command line when running the executable of "CCX" with the adapter being used (this is described later). Also, the name must be the same as the one used in the preCICE configuration file *precice-config.xml*.  

Reviewer: ```suggestion
The adapter allows us to use several participants in one simulation (e.g., several instances of Calculix if several solid objects are taken into account). The name of the participant `Calculix` must match the specification of the participant on the command line when running the executable of `CCX` with the adapter being used (this is described later). Also, the name must be the same as the one used in the preCICE configuration file `precice-config.xml`.  
```

Depending on the data you need to read and write, the interface should define either a `faces-mesh` (or simply `mesh` as a synonym) where the data points are centers of faces (computed by the adapter) or a mesh made of CalculiX vertices, with the keyword `nodes-mesh`. An interface made of faces should be defined in the CalculiX case using the `*SURFACE` command, whereas meshes with nodes should define these nodes using `*NSET`. Using the wrong family of mesh (e.g. reading forces on faces) throws an error. If you need both kinds of meshes, you should define more than one interface.

For FSI simulations the mesh type of an interface is always `nodes-mesh`, as forces and displacement are defined on nodes. The name of this mesh, `Calculix_Mesh`, must match the mesh name given in the preCICE configuration file. In CHT simulations, `faces-meshes` are usually chosen, as they are needed to apply heat fluxes or convective heat transfer.
For defining which nodes of the CalculiX domain belong to the FSI interface, a node set needs to be defined in the CalculiX input files. The name of this node set must match the name of the patch (here: "interface").  

> In the current FSI example, the adapter reads forces from preCICE and feeds displacement deltas (not absolute displacements, but the change of the displacements relative to the last time step) to preCICE. This is defined with the keywords "read-data" and "write-data", respectively. The names (here: "Forces" and "DisplacementDeltas") again need to match the specifications in the preCICE configuration file. In the current example, the coupled fluid solver expects displacement deltas instead of displacements. However, the adapter is capable of writing either type. Just use "write-data: [Displacements]" for absolute displacements rather than relative changes being transferred in each time step. Valid `readData` keywords in CalculiX are:

Reviewer: ```suggestion
In the current FSI example, the adapter reads forces from preCICE and feeds displacement deltas (not absolute displacements, but the change of the displacements relative to the last time step) to preCICE. This is defined with the keywords `read-data` and `write-data`, respectively. The names (here: `Forces` and `DisplacementDeltas`) again need to match the specifications in the preCICE configuration file. In the current example, the coupled fluid solver expects displacement deltas instead of displacements. However, the adapter is capable of writing either type. Just use `write-data: [Displacements]` for absolute displacements rather than relative changes being transferred in each time step. Valid `readData` keywords in CalculiX are:
```

```

The adapter allows to use several participants in one simulation (e.g. several instances of Calculix if several solid objects are taken into account). The name of the participant "Calculix" must match the specification of the participant on the command line when running the executable of "CCX" with the adapter being used (this is described later). Also, the name must be the same as the one used in the preCICE configuration file *precice-config.xml*.  
One participant may have several coupling interfaces. Note that each interface specification starts with a dash.

> Depending on the data you need to read and write, you can use a mesh made of face centers ("faces-mesh" or just "mesh") or a mesh made of actual CalculiX nodes.

Reviewer: What does `("faces-mesh" or just "mesh")` mean? Is this something important for some configuration, or just a forward declaration for this text (like in legal texts)?

Also, why "actual CalculiX nodes" but just "face centers"? If this is important, we should probably explain that the adapter defines face centers by itself (I don't know how it is implemented).

Author: You decide whether you use a faces-mesh or a nodes-mesh (depending only on the kin of data you read/write) and declare it. If you declare a faces-mesh then configure "reading forces" you should get an error (or a crash ? Error handling is not optimal yet).
"mesh" is a synonym of "faces-mesh" (probably legacy where the original adapter used almost exclusively faces-mesh ?)

Face centers are computed by the adapter as averages of the nodes positions defining them

Reviewer: Nice explanation, let's document it in the actual documentation! :-)

> For FSI simulations the mesh type of an interface is always "nodes-mesh", as forces and displacement are defined on nodes. The name of this mesh, "Calculix_Mesh", must match the mesh name given in the preCICE configuration file. In CHT simulations, faces mesh are usually chosen, as they are needed to apply heat fluxes or convective heat transfer.

Reviewer: Consistency: 
```suggestion
For FSI simulations the mesh type of an interface is always "nodes-mesh", as forces and displacements are defined on nodes. The name of this mesh, "Calculix_Mesh", must match the mesh name given in the preCICE configuration file. In CHT simulations, "faces-mesh" are usually chosen, as they are needed to apply heat fluxes or convective heat transfer.
```

(even though I would prefer `node-based meshes` and `face-based meshes`)

Author: The convention seems to favor keywords instead of natural language.

For defining which nodes of the CalculiX domain belong to the FSI interface, a node set needs to be defined in the CalculiX input files. The name of this node set must match the name of the patch (here: "interface").  
In the current FSI example, the adapter reads forces from preCICE and feeds displacement deltas (not absolute displacements, but the change of the displacements relative to the last time step) to preCICE. This is defined with the keywords "read-data" and "write-data", respectively. The names (here: "Forces" and "DisplacementDeltas") again need to match the specifications in the preCICE configuration file. In the current example, the coupled fluid solver expects displacement deltas instead of displacements. However, the adapter is capable of writing either type. Just use "write-data: [Displacements]" for absolute displacements rather than relative changes being transferred in each time step. Valid `readData` keywords in CalculiX are:

```text
On faces-mesh:

> * Pressure (Apply a *DLOAD)

Reviewer: Where should there `*DLOAD`, `*DFLUX`, `*FILM` etc be applied? One opening sentence explaining this (ideally linking to the CalculiX documentation) would help.
Also a minor consistency thing: `Apply a` vs `Use`.

Author: The only up-to-date documentation is either a 700 pages PDF or a downloadable HTML repository. Accurate links are thus quite hard to write.
Someone uploaded them on the web (https://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node188.html) but it's for version 2.7, which is quite old. Not sure what the optimal solution is.

Again, I'm careful to avoid making our documentation a subset of CalculiX documentation, but on the other hand it's true that explanations are scarce.

Reviewer: Then at least let's write which file this concerns. We have a similar link for the element types, which is very helpful. If only an old version is available online, that's also fine. I guess anything 2.x is going to be fine, CalculiX is not such a quickly moving target. We can even just refer to sections (by number + name, as this may change).

Regarding the balance: I always think of our users as people that don't really know CalculiX, but it is the most common choice for the solid part of FSI simulations in our community. So, if we help people understand CalculiX, we help them successfully use preCICE in their project. I think a few hints are always welcome, and that's what I would expect here ("`*DLOAD` for blah, `*DFLUX` for blah, ..."), at least if linking to the CalculiX docs is not easy.

* Heat-Flux (Apply a *DFLUX)
* Sink-Temperature (Use *FILM)
* Heat-Transfer-Coefficient (Use *FILM)
On nodes-mesh:
* Forces (Apply a *CLOAD)
* Displacements (Use *BOUNDARY)
* Temperature (Use *BOUNDARY)

> ```

Reviewer: Why do we actually have this in a code block? Better convert to two lists with inline code where needed (same below).

```

The adapter allows to use several participants in one simulation (e.g. several instances of Calculix if several solid objects are taken into account). The name of the participant "Calculix" must match the specification of the participant on the command line when running the executable of "CCX" with the adapter being used (this is described later). Also, the name must be the same as the one used in the preCICE configuration file *precice-config.xml*.  

One participant may have several coupling interfaces. Note that each interface specification starts with a dash.

> Depending on the data you need to read and write, the interface should define either a "faces-mesh" (or simply "mesh" as a synonym) where the data points are centers of faces (computed by the adapter) or a mesh made CalculiX vertices, with the keyword "nodes-mesh". An interface made of faces should be defined in the CalculiX case using the `*SURFACE` command, whereas meshes with nodes should define these nodes using `*NSET`. Using the wrong family of mesh (e.g. reading forces on faces) throws an error. If you need both kinds of meshes, you should define more than one interface.

Reviewer: Only looking at the diff, I was a bit confused if `"faces-mesh"` was a code keyword or an English keyword. Typesetting as code makes this crystal clear.

```suggestion
Depending on the data you need to read and write, the interface should define either a `faces-mesh` (or simply `mesh` as a synonym) where the data points are centers of faces (computed by the adapter) or a mesh made of CalculiX vertices, with the keyword `nodes-mesh`. An interface made of faces should be defined in the CalculiX case using the `*SURFACE` command, whereas meshes with nodes should define these nodes using `*NSET`. Using the wrong family of mesh (e.g. reading forces on faces) throws an error. If you need both kinds of meshes, you should define more than one interface.
```

> For FSI simulations the mesh type of an interface is always "nodes-mesh", as forces and displacement are defined on nodes. The name of this mesh, "Calculix_Mesh", must match the mesh name given in the preCICE configuration file. In CHT simulations, "faces-meshes" are usually chosen, as they are needed to apply heat fluxes or convective heat transfer.

Reviewer: ```suggestion
For FSI simulations the mesh type of an interface is always `nodes-mesh`, as forces and displacement are defined on nodes. The name of this mesh, `Calculix_Mesh`, must match the mesh name given in the preCICE configuration file. In CHT simulations, `faces-meshes` are usually chosen, as they are needed to apply heat fluxes or convective heat transfer.
```

For defining which nodes of the CalculiX domain belong to the FSI interface, a node set needs to be defined in the CalculiX input files. The name of this node set must match the name of the patch (here: "interface").  

In the current FSI example, the adapter reads forces from preCICE and feeds displacement deltas (not absolute displacements, but the change of the displacements relative to the last time step) to preCICE. This is defined with the keywords "read-data" and "write-data", respectively. The names (here: "Forces" and "DisplacementDeltas") again need to match the specifications in the preCICE configuration file. In the current example, the coupled fluid solver expects displacement deltas instead of displacements. However, the adapter is capable of writing either type. Just use "write-data: [Displacements]" for absolute displacements rather than relative changes being transferred in each time step. Valid `readData` keywords in CalculiX are:

On faces-mesh:

> * Pressure (Use a `*DLOAD`)

Reviewer: Fixing a Markdown linter error:

```suggestion

* Pressure (Use a `*DLOAD`)
```

* Heat-Flux (Use a `*DFLUX`)
* Sink-Temperature (Use `*FILM`)

> * Heat-Transfer-Coefficient (Use `*FILM`)
> On nodes-mesh:
> * Forces (Use a `*CLOAD`)

Reviewer: ```suggestion
* Heat-Transfer-Coefficient (Use `*FILM`)

On nodes-mesh:

* Forces (Use a `*CLOAD`)
```

* Displacements (Use `*BOUNDARY`)
* Temperature (Use `*BOUNDARY`)

> Have a look at the CalculiX documentation for a detailed description of each of these commands. There is an [online (but outdated) version ](https://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node1.html) and an [up-to-date PDF version](http://www.dhondt.de/ccx_2.19.pdf).

Reviewer: Fixing a Markdown linter error:

```suggestion
Have a look at the CalculiX documentation for a detailed description of each of these commands. There is an [online (but outdated) version](https://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node1.html) and an [up-to-date PDF version](http://www.dhondt.de/ccx_2.19.pdf).
```

 Valid `writeData` keywords are:

On faces-mesh:
* Pressure
* Heat-Flux
* Sink-Temperature

> * Heat-Transfer-Coefficient
> On nodes-mesh:
> * Forces

Reviewer: ```suggestion
* Heat-Transfer-Coefficient

On nodes-mesh:

* Forces
```

The adapter internally uses the CalculiX data format for point forces to apply the FSI forces at the coupling interface. This data structure is only initialized for those nodes, which are loaded at the beginning of a CalculiX analysis step via the input file. Thus, it is necessary to load all nodes of the node set, which defines the FSI interface in CalculiX (referring to the above example, the nodes of set "interface" (Note that in CalculiX a node set always begins with an "N" followed by the actual name of the set, which is here "interface".) are loaded via the "CLOAD" keyword.), in each spatial direction. However, the values of these initial forces can (and should) be chosen to zero, such that the simulation result is not affected.

> When using faces mesh, instead of a node set (\*NSET), a \*SURFACE must be sent, defined by a list of elements and face numbers. Instead of starting with a "N", the name must start with a "S".

Reviewer: ```suggestion
When using faces mesh, instead of a node set (`*NSET`), a `*SURFACE` must be sent, defined by a list of elements and face numbers. Instead of starting with a `N`, the name must start with a `S`. For example: `SCalculiX_Mesh` (while in the preCICE configuration the name remains `CalculiX_Mesh`).
```

The adapter internally uses the CalculiX data format for point forces to apply the FSI forces at the coupling interface. This data structure is only initialized for those nodes, which are loaded at the beginning of a CalculiX analysis step via the input file. Thus, it is necessary to load all nodes of the node set, which defines the FSI interface in CalculiX (referring to the above example, the nodes of set "interface" (Note that in CalculiX a node set always begins with an "N" followed by the actual name of the set, which is here "interface".) are loaded via the "CLOAD" keyword.), in each spatial direction. However, the values of these initial forces can (and should) be chosen to zero, such that the simulation result is not affected.

> When using "faces-meshes", instead of a node set (\*NSET), a \*SURFACE must be sent, defined by a list of elements and face numbers. Instead of starting with a "N", the name must start with a "S".

Reviewer: ```suggestion
When using "faces-meshes", instead of a node set (`*NSET`), a `*SURFACE` must be sent, defined by a list of elements and face numbers. Instead of starting with a `N`, the name must start with a `S`.
```

Reviewer: Still open

CalculiX CCX offers both a geometrically linear as well as a geometrically non-linear solver. Both are coupled via the adapter. The keyword "NLGEOM" (as shown in the example) needs to be included in the CalculiX case input file in order to select the geometrically non-linear solver. It is also automatically triggered if material non-linearities are included in the analysis. In case the keyword "NLGEOM" does not appear in the CalculiX case input file and the chosen materials are linear, the geometrically linear CalculiX solver is used. In any case, for FSI simulations via preCICE the keyword "DYNAMIC" (enabling a dynamic computation) must appear in the CalculiX input file.

More input files that you may find in the CalculiX tutorial cases:

> t

Reviewer: Stray

```suggestion

```

### Supported elements

> The preCICE CalculiX adapter should support most elements when using nodes meshes. It has been used with both linear and quadratic tetrahedral (C3D4 and C3D10) and hexahedral (C3D8, C3D8I, and [C3D20](http://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node29.html)) elements. There is however a restriction when using nearest-projection mapping: in that case, you have to use tetrahedral elements.

Reviewer: ```suggestion
The preCICE CalculiX adapter should support most elements when using nodes meshes. It has been used with both linear and quadratic tetrahedral (C3D4 and C3D10) and hexahedral (C3D8, C3D8I, and [C3D20](http://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node29.html)) elements. There is, however, a restriction when using nearest-projection mapping: in that case, you have to use tetrahedral elements.
```

When using face meshes, only tetrahedra and hexaedra are supported.

### Coupling to 2D simulations

> The adapter supports quasi 2D simulations when the z-direction is ignored. If you set the preCICE interface dimension to 2, the adapter will map data from the CalculiX 3D simulation to 2D space and vice-versa. The 3D simulation should be made of solid elements (or shells) of unit thickness.

Reviewer: ```suggestion
The adapter supports quasi-2D simulations when the z-direction is ignored. If you set the preCICE interface dimension to 2, the adapter will map data from the CalculiX 3D simulation to the 2D space and vice-versa. The 3D simulation should be made of solid elements (or shells) of unit thickness.
```

#### Behavior with nodes meshes

> When writing continuous fields (such as temperature and displacements), the adapter will send data that is averaged over thickness. For conservative data (such as forces), sums are computed. When reading forces, the load applied to a 2D point will be spread evenly between the 3D points sharing the same x and y coordinates.

Reviewer: Either singular everywhere, or plural everywhere:
```suggestion
When writing continuous fields (such as temperature and displacement), the adapter will send data that is averaged over thickness. For conservative data (such as forces), sums are computed. When reading forces, the load applied to a 2D point will be spread evenly between the 3D points sharing the same x and y coordinates.
```

> #### Behavior with faces meshes
> 
> When using faces mesh, the behavior in unchanged and the z-component is simply discarded.

Reviewer: ```suggestion
#### Behavior with face-based meshes

When using face-based mesh, the behavior in unchanged and the z-component is simply discarded.
```

Author: Shouldn't we use the keywirds (`"faces-mesh"`) for consistency ?

Reviewer: Anything, just the same everywhere. Even though I find it a bit strange using code keywords in headings (but that's just my gut feeling).

### Supported elements

> The preCICE CalculiX adapter should support most elements when using "nodes-meshes". It has been used with both linear and quadratic tetrahedral (C3D4 and C3D10) and hexahedral (C3D8, C3D8I, and [C3D20](http://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node29.html)) elements. There is however a restriction when using nearest-projection mapping: in that case, you have to use tetrahedral elements.

Reviewer: ```suggestion
The preCICE CalculiX adapter should support most elements when using `nodes-meshes`. It has been used with both linear and quadratic tetrahedral (C3D4 and C3D10) and hexahedral (C3D8, C3D8I, and [C3D20](http://web.mit.edu/calculix_v2.7/CalculiX/ccx_2.7/doc/ccx/node29.html)) elements. There is however a restriction when using nearest-projection mapping: in that case, you have to use tetrahedral elements.
```

When using face meshes, only tetrahedra and hexaedra are supported.

### Coupling to 2D simulations

The adapter supports quasi 2D simulations when the z-direction is ignored. If you set the preCICE interface dimension to 2, the adapter will map data from the CalculiX 3D simulation to 2D space and vice-versa. The 3D simulation should be made of solid elements (or shells) of unit thickness.

> #### Behavior with "nodes-meshes"

Reviewer: ```suggestion
#### Behavior with `nodes-mesh`
```

When writing continuous fields (such as temperature and displacements), the adapter will send data that is averaged over thickness. For conservative data (such as forces), sums are computed. When reading forces, the load applied to a 2D point will be spread evenly between the 3D points sharing the same x and y coordinates.

> #### Behavior with "faces-meshes"

Reviewer: ```suggestion
#### Behavior with `faces-mesh`
```

> When using "faces-meshes", the behavior in unchanged and the z-component is simply discarded.

Reviewer: ```suggestion
When using `faces-mesh`, the behavior in unchanged and the z-component is simply discarded.
```

Note that an error will only occur if nodes-mesh-with-connectivity is specified without a .sur file. The calculix-adapter with nearest-projection mapping only supports tetrahedral elements (C3D4 and C3D10) as preCICE only works with surface triangles for nearest-projection mapping.

### Modal dynamic simulations

> The adapter supports modal dynamic simulations, where eigenmodes from a frequency analysis are used. Run the frequency analysis without preCICE (`ccx_preCICE -i [CalculiX input file]`) then run a coupled simulation. Currently, only explicit coupling is supported.

Reviewer: Half a sentence more explaining the concept behind modal dynamic simulations would be appreciated.

Note that an error will only occur if nodes-mesh-with-connectivity is specified without a .sur file. The calculix-adapter with nearest-projection mapping only supports tetrahedral elements (C3D4 and C3D10) as preCICE only works with surface triangles for nearest-projection mapping.

### Modal dynamic simulations

> The adapter supports modal dynamic simulations, where eigenmodes from a frequency analysis are used. Instead of solving the full system of equations, CalculiX solves the problem as a linear combination of osicllating eigenmodes. This can be faster but slightly less accurate. It can be particularly useful if you need many runs of the simulation, as frequency analysis can be run only once. Run the frequency analysis without preCICE (`ccx_preCICE -i [CalculiX input file]`) then run a regular coupled simulation (configured as modal dynamic in the input file). Note that reading displacements is not supported in these simulations.

Reviewer: ```suggestion
The adapter supports modal dynamic simulations, where eigenmodes from a frequency analysis are used. Instead of solving the full system of equations, CalculiX solves the problem as a linear combination of osicllating eigenmodes. This can be faster but slightly less accurate. It can be particularly useful if you need many runs of the simulation, as frequency analysis can be run only once. Run the frequency analysis without preCICE (`ccx_preCICE -i [CalculiX input file]`), then run a regular coupled simulation (configured as modal dynamic in the input file). Note that reading displacements is not supported in these simulations.
```

(nice additions!)
